### PR TITLE
feat: incoming request validation

### PR DESCRIPTION
**Pull request by grace0925**
Closes: #6 
New PR after accidentally closing PR #126 
Signed-off-by: Grace Liu <grace.liu@securekey.com>

### CODE REVIEW

File: pkg/edvutils/edvutils.go [hunk 1]
```diff
@@ -63,3 +71,89 @@ func Base58Encoded128BitToUUID(name string) (string, error) {
...
+		if err != nil {
+			return err
+		}
+
+		if foundAlg {
+			return nil
+		}
+	}
+
+	if jwe.SingleRecipientHeader != nil && jwe.SingleRecipientHeader.Alg != "" {
```
Comment (llorllale) [inline pkg/edvutils/edvutils.go:118]: we should also ensure its value is not `none`

**Review by llorllale — APPROVED** [review]

File: pkg/edvutils/edvutils.go (hunk 1 above)
Comment (grace0925) [reply pkg/edvutils/edvutils.go]: fixed

File: pkg/edvutils/edvutils.go [hunk 2]
```diff
@@ -63,3 +74,98 @@ func Base58Encoded128BitToUUID(name string) (string, error) {
...
+
+	if err := checkAlg(&jwe); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkAlg(jwe *models.JSONWebEncryption) error {
+	if jwe.B64ProtectedHeaders != "" && jwe.B64ProtectedHeaders != none {
```
Comment (DRK3) [inline pkg/edvutils/edvutils.go:118]: Don't you need to decode the B64ProtectedHeaders _before_ you check for a `none` alg?

File: pkg/edvutils/edvutils.go [hunk 3]
```diff
@@ -63,3 +74,98 @@ func Base58Encoded128BitToUUID(name string) (string, error) {
...
+		foundAlg, err := checkAlgInProtectedHeader(jwe.B64ProtectedHeaders)
+		if err != nil {
+			return err
+		}
+
+		if foundAlg {
+			return nil
```
Comment (DRK3) [inline pkg/edvutils/edvutils.go:125]: It looks like you're only checking to see if `alg` exists - but don't you also need to make sure that it's not `none`?

File: pkg/edvutils/edvutils.go [hunk 4]
```diff
@@ -63,3 +74,98 @@ func Base58Encoded128BitToUUID(name string) (string, error) {
...
+		if err := CheckIfURI(str); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateRawJWE returns an error if the given raw JWE is empty or has invalid alg fields.
+func ValidateRawJWE(rawJWE json.RawMessage) error {
```
Comment (DRK3) [inline pkg/edvutils/edvutils.go:100]: It's not "Raw JWE" - it's just JWE. a json.RawMessage is just a []byte that behaves differently when serialized.

You can call this method just "ValidateJWE".

(You might even consider seeing if the method signature can just use a []byte directly. Not sure if this is really a better thing to do, but try it out and see)

File: pkg/edvutils/edvutils.go (hunk 2 above)
Comment (grace0925) [reply pkg/edvutils/edvutils.go]: fixed

File: pkg/edvutils/edvutils.go (hunk 3 above)
Comment (grace0925) [reply pkg/edvutils/edvutils.go:125]: fixed

File: pkg/edvutils/edvutils.go (hunk 4 above)
Comment (grace0925) [reply pkg/edvutils/edvutils.go]: I changed it to using []byte directly

File: pkg/edvutils/edvutils.go (hunk 3 above)
Comment (DRK3) [reply pkg/edvutils/edvutils.go:125]: It looks like you're still not checking to see if `alg` is `none` for the protected headers

Comment (DRK3) [reply pkg/edvutils/edvutils.go:125]: Oops, my bad - I missed it

**Review by DRK3 — APPROVED** [review]

**Pull request merged**